Add user interface for graph editor graph layout parameters

thomaskrause opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help. -->

## Pull request type

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. 
Choose one of the following types (you can copy and paste them if you like)

- Bugfix
- Feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no API changes)
- Build related changes
- Documentation content changes
- Other (please describe it)

--> 

Feature

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: closes #133 


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

- It is possible to configure the different margins in the graph editor view in the UI
  - Vertical margin between all nodes
  - Horizontal margin between the token nodes (affects also spacing between the ancestor nodes)
  - Additional margin between the token row and the other nodes, useful for adding space for depency edges 
- Restructured the user documentation for the graph editor because there is now more optional functionality that should not all be described on the first page.
- Moving (Swt)ResourceManager classes from the WindowBuilder into core bundle, since they are used by at least two bundles to manage the icons. Otherwise, we would have to duplicate these two classes. Since these classes are now in core, I adapted them to our code style and fixed multiple code smells. Also, on the different places where font and image resources are already created, the new manger classes are now used.



## Does this introduce a breaking change?

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

N/A

## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

![Peek 2022-09-08 18-00](https://user-images.githubusercontent.com/2168104/189170082-12cc9282-beca-4f6e-b3d7-1da52cbd1957.gif)



## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added **or** the PR is neither a bug fix nor a feature
- [x] Docs have been reviewed and added / updated if needed **or** the PR is neither a bug fix nor a feature
- [x] The *Unreleased* section in CHANGELOG.md has been amended to reflect the changes in this PR if needed
- [x] Build (`mvn -Pcff verify`) was run locally and any changes were pushed
- [x] The pull request is against the correct branch (`main` for bug fixes, `develop` for new functionality)
- [x] [Dependencies and citation templates](https://github.com/hexatomic/hexatomic/tree/develop/releng/templates) have been updated where necessary **or** there are no new dependencies

---

# Checklist for maintainers

## Releases

- Do **NOT** push the green merge button on GitHub.  
Follow the latest version of the [developer/maintainer docs](https://hexatomic.github.io/hexatomic/dev/) for making releases.
- Check that license and citation information for dependencies are complete (each dependency has a folder in `/THIRD-PARTY/`).

sdruskat commented on docs/user/src/usage/graph-editor/README.md:
```diff
@@ -48,9 +37,5 @@ You can navigate the graph view as follows:
-If you don't like the layout of the graph, you can change it by simply dragging nodes with your mouse.
-
-## Console
+If you don't like the layout of the graph, you can change it by simply dragging nodes with your mouse or [adjust the layout parameters](./layout-params.md).
```
```suggestion
If you don't like the layout of the graph, you can change it by dragging nodes with your mouse or [adjusting the layout parameters](./layout-params.md).
```

Remove documentation smell, fix grammar.

sdruskat commented on docs/user/src/usage/graph-editor/README.md:
```diff
@@ -5,28 +5,17 @@ It provides a general visualization that displays all possible types of annotati
 ![Screenshot of the graph editor](graph-viewer.png)
 
-## Filters
+On the bottom of the graph editor, there is the console which you can use to actually edit the graph.
```
```suggestion
On the bottom of the graph editor is the console that you can use to edit the graph.
```

Simply sentence.

sdruskat commented on docs/user/src/usage/graph-editor/filter.md:
```diff
@@ -0,0 +1,17 @@
+
+![Screenshot of the graph editor with expanded filter view](filter-view.png)
+
+You can choose to display spans and their annotations in the graph, by first expanding **Annotation Types** and the checking the checkbox **Spans**. 
```
```suggestion
You can choose to display spans and their annotations in the graph, by first expanding **Annotation Types** and then checking the checkbox **Spans**. 
```

Fix typo.

sdruskat commented on docs/user/src/usage/graph-editor/filter.md:
```diff
@@ -0,0 +1,17 @@
+![Screenshot of the graph editor with expanded filter view](filter-view.png)
+
+You can choose to display spans and their annotations in the graph, by first expanding **Annotation Types** and the checking the checkbox **Spans**. 
+Spans are special nodes to collect a number of tokens and to annotate them all at once.
```
```suggestion
Spans are special nodes that "collect" a number of tokens.
Span annotations can then be made for that *specific set* of tokens, e.g., a phrase, clause or sentence.
```

Improve wording. "Annotating all at once" sounds like you would use spans to add annotations to all tokens, whereas spans in itself are meaningful units.

sdruskat commented on docs/user/src/usage/graph-editor/filter.md:
```diff
@@ -0,0 +1,17 @@
+You can choose to display spans and their annotations in the graph, by first expanding **Annotation Types** and the checking the checkbox **Spans**. 
+Spans are special nodes to collect a number of tokens and to annotate them all at once.
+If you want to learn more about spans, please read the [Salt documentation](http://corpus-tools.org/salt/#documentation).
+Similarly, you can show or hide pointing relations between nodes in the graph by using the checkbox **Pointing Relations**.
```
```suggestion

Similarly, you can show or hide pointing relations between nodes in the graph by using the checkbox **Pointing Relations**.
```

Add blank line to separate this point from the span explanation above.

sdruskat commented on docs/user/src/usage/graph-editor/filter.md:
```diff
@@ -0,0 +1,17 @@
+
+**Node Annotations** allows you to filter the segments that include annotations of a specific *name*.
+To add a filter criterion (also called a "facet"), search for an annotation name in the text field and select the matching annotation.
+This creates a box with the annotation name of the applied filter.
```
```suggestion
This creates a new filter badge with the annotation name of the applied filter.
```

I think "filter badge" is a better name for the thing.

sdruskat commented on docs/user/src/usage/graph-editor/filter.md:
```diff
@@ -0,0 +1,17 @@
+**Node Annotations** allows you to filter the segments that include annotations of a specific *name*.
+To add a filter criterion (also called a "facet"), search for an annotation name in the text field and select the matching annotation.
+This creates a box with the annotation name of the applied filter.
+You can add more than one filter and any node, that contains any of the selected annotation names, will be shown.
```
```suggestion
You can add more than one filter. All nodes that contain any of the selected annotation names will be shown.
```

Separate sentences.

sdruskat commented on docs/user/src/usage/graph-editor/filter.md:
```diff
@@ -0,0 +1,17 @@
+To add a filter criterion (also called a "facet"), search for an annotation name in the text field and select the matching annotation.
+This creates a box with the annotation name of the applied filter.
+You can add more than one filter and any node, that contains any of the selected annotation names, will be shown.
+Click on the close button to delete the annotation name filter again.
```
```suggestion
Click on the close button on a filter badge to delete the respective annotation name filter.
```

Make more precise.

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+# Change layout parameters
+
+Depending on the shown annotations and the specific graph, the default visualization of the graph might get cluttered.
```
```suggestion
Depending on the visible annotations and the specific graph, the default visualization of the graph might be cluttered.
```

Improve wording.

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+# Change layout parameters
+
+Depending on the shown annotations and the specific graph, the default visualization of the graph might get cluttered.
+The graph editor will try to arrange the nodes using a default layout algorithm, which parameters you can adjust by expanding the "Graph Layout" section.
```
```suggestion
The graph editor will try to arrange the nodes using a default layout algorithm.
You can adjust the layout parameters by expanding the "Graph Layout" section.
```

Separate sentences.

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+
+![Graph editor view with expanded layout parameters](./layout-params.png)
+
+## Horizontal margin between token
```
```suggestion
## Horizontal margin between tokens
```

Use plural

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+![Horizontal margin parameter slider](./layout-param-horizontal-margin.png)
+
+
+There is some horizontal space between the token at the bottom of the graph.
```
```suggestion
This parameter changes the horizontal space between the tokens at the bottom of the graph.
```

Improve wording.

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+There is some horizontal space between the token at the bottom of the graph.
+If the space is increased, this indirectly affects the space between other nodes, too.
+
+This is measured in "times of the average token width".
```
```suggestion
This parameter is measured in "multiples of the average token width".
```

Improve wording.

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+If the space is increased, this indirectly affects the space between other nodes, too.
+
+This is measured in "times of the average token width".
+So for \"0\" there is no margin, for "1" the margin has  the same width as the average token node, and for "2" the margin is twice as high as the average token node width.
```
```suggestion
So for \"0\" there is no margin, for "1" the margin has the same width as the average token node, and for "2" the margin is twice as high as the average token node width.
```

Remove duplicate whitespace.

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+
+![Vertical margin parameter slider](./layout-param-vertical-margin.png)
+
+Configures the vertical margin between *all* nodes.
```
```suggestion
This parameter configures the vertical margin between *all* nodes.
```

Make this a sentence.

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+
+Configures the vertical margin between *all* nodes.
+
+This is measure in "times of the node height".
```
```suggestion
This is measured in "multiples of the node height".
```

Fix typo, wording.

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+![Effect of the settings 0, 1 and 2 for the vertical margin parameter](vertical-margin-examples.png)
+
+
+## Vertical margin between token and non-token
```
```suggestion
## Vertical margin between tokens and non-tokens
```

Plural.

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+![Vertical margin between token and non-token parameter slider](./layout-param-token-margin.png)
+
+
+Token are grouped in the lowest rank (at the bottom)
```
```suggestion
Tokens are grouped horizontally at the bottom of the graph visualization, i.e., in the bottom "row".
```

Introduce row before it is used later, improve wording.

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+
+
+Token are grouped in the lowest rank (at the bottom)
+To allow space for pointing relations, you can add a margin between the token row and the lowest annotation nodes.
```
```suggestion
To allow space for pointing relations, you can add a margin between the token "row" and the annotation nodes in the "row" just above it.
```

sdruskat commented on docs/user/src/usage/graph-editor/layout-params.md:
```diff
@@ -0,0 +1,45 @@
+To allow space for pointing relations, you can add a margin between the token row and the lowest annotation nodes.
+This margin is in addition to the [vertical margin between nodes](#vertical-margin-between-nodes).
+
+A margin of "1" means there is one level left empty, an offset of "0"
```
```suggestion
A margin of "1" means there is one empty level added between the token "row" and the annotation node "row" above it, an offset of "0"
```

Make clearer.

sdruskat reviewed (approved):
Thanks! Great stuff.

I've added a number of wording documentation improvement suggestions. Otherwise this is good to go.

One thing I realized is that when changing horizontal spacing is that the current focus is lost, i.e., the view scrolls to a different section of the graph. Whether fixing this is in scope for this PR is up to you. Could be a new PR just as well (in which case feel free to turn this into an issue).

![hex](https://user-images.githubusercontent.com/3007126/191238834-c203a292-8691-41ec-b3a3-7ca3ccf4ca60.gif)

thomaskrause commented:
> Thanks! Great stuff.
> 
> I've added a number of wording documentation improvement suggestions. Otherwise this is good to go.
> 
> One thing I realized is that when changing horizontal spacing is that the current focus is lost, i.e., the view scrolls to a different section of the graph. Whether fixing this is in scope for this PR is up to you. Could be a new PR just as well (in which case feel free to turn this into an issue).
> 
> ![hex](https://user-images.githubusercontent.com/3007126/191238834-c203a292-8691-41ec-b3a3-7ca3ccf4ca60.gif) [ ![hex](https://user-images.githubusercontent.com/3007126/191238834-c203a292-8691-41ec-b3a3-7ca3ccf4ca60.gif) ](https://user-images.githubusercontent.com/3007126/191238834-c203a292-8691-41ec-b3a3-7ca3ccf4ca60.gif) [ ](https://user-images.githubusercontent.com/3007126/191238834-c203a292-8691-41ec-b3a3-7ca3ccf4ca60.gif)

I added an issue for this, since it seems to work under Linux with an example corpus for me (I have trouble creating GIF screenshots, but will add the example later). So this could be related to the specific example corpus or to the Windows platform.
